From 228b80f449453fdfba5cf558e618d23da8e44ca0 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Fri, 29 Nov 2019 18:14:07 +0100
Subject: [PATCH] PONRTSYS-5630: Add Linux API for MAC table limit

This allows to configure the maximum MAC addressees which can be learned
on a bridge port. This is currently not implemented in software only the
hardware offloading part is working.

These settings are per port and can be different for each port in a
bridge.

-1 means that the port allows unlimited learning which is also the
default setting.
0 means no learning should be possible on this port.
---
 include/net/switchdev.h      |  2 ++
 include/uapi/linux/if_link.h |  3 +++
 net/bridge/br_if.c           |  1 +
 net/bridge/br_netlink.c      | 44 +++++++++++++++++++++++++++++++++++++++++++-
 net/bridge/br_private.h      |  1 +
 net/bridge/br_sysfs_if.c     |  7 +++++++
 6 files changed, 57 insertions(+), 1 deletion(-)

diff --git a/include/net/switchdev.h b/include/net/switchdev.h
index eba80c4fc56f..2fa2cbf7c10e 100644
--- a/include/net/switchdev.h
+++ b/include/net/switchdev.h
@@ -46,6 +46,7 @@ enum switchdev_attr_id {
 	SWITCHDEV_ATTR_ID_PORT_PARENT_ID,
 	SWITCHDEV_ATTR_ID_PORT_STP_STATE,
 	SWITCHDEV_ATTR_ID_PORT_BRIDGE_FLAGS,
+	SWITCHDEV_ATTR_ID_PORT_LEARNING_LIMIT,
 	SWITCHDEV_ATTR_ID_BRIDGE_AGEING_TIME,
 	SWITCHDEV_ATTR_ID_BRIDGE_VLAN_FILTERING,
 };
@@ -60,6 +61,7 @@ struct switchdev_attr {
 		struct netdev_phys_item_id ppid;	/* PORT_PARENT_ID */
 		u8 stp_state;				/* PORT_STP_STATE */
 		unsigned long brport_flags;		/* PORT_BRIDGE_FLAGS */
+		int learning_limit;			/* PORT_LEARNING_LIMIT */
 		clock_t ageing_time;			/* BRIDGE_AGEING_TIME */
 		bool vlan_filtering;			/* BRIDGE_VLAN_FILTERING */
 	} u;
diff --git a/include/uapi/linux/if_link.h b/include/uapi/linux/if_link.h
index 89389d8403ad..4d696b6b3815 100644
--- a/include/uapi/linux/if_link.h
+++ b/include/uapi/linux/if_link.h
@@ -290,6 +290,8 @@ enum {
 	BRIDGE_MODE_HAIRPIN,
 };
 
+#define BRPORT_LEARNING_LIMIT_UNLIMITED -1
+
 enum {
 	IFLA_BRPORT_UNSPEC,
 	IFLA_BRPORT_STATE,	/* Spanning tree state     */
@@ -326,6 +328,7 @@ enum {
 	IFLA_BRPORT_NEIGH_SUPPRESS,
 	IFLA_BRPORT_ISOLATED,
 	IFLA_BRPORT_BACKUP_PORT,
+	IFLA_BRPORT_LEARNING_LIMIT,
 	__IFLA_BRPORT_MAX
 };
 #define IFLA_BRPORT_MAX (__IFLA_BRPORT_MAX - 1)
diff --git a/net/bridge/br_if.c b/net/bridge/br_if.c
index 0f66d7ee642e..5021bba8e8f4 100644
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -364,6 +364,7 @@ static struct net_bridge_port *new_nbp(struct net_bridge *br,
 	dev_hold(dev);
 	p->dev = dev;
 	p->path_cost = port_cost(dev);
+	p->learning_limit = BRPORT_LEARNING_LIMIT_UNLIMITED;
 	p->priority = 0x8000 >> BR_PORT_BITS;
 	p->port_no = index;
 	p->flags = BR_LEARNING | BR_FLOOD | BR_MCAST_FLOOD;
diff --git a/net/bridge/br_netlink.c b/net/bridge/br_netlink.c
index 4f831225d34f..72191e0d4914 100644
--- a/net/bridge/br_netlink.c
+++ b/net/bridge/br_netlink.c
@@ -16,6 +16,7 @@
 #include <net/rtnetlink.h>
 #include <net/net_namespace.h>
 #include <net/sock.h>
+#include <net/switchdev.h>
 #include <uapi/linux/if_bridge.h>
 
 #include "br_private.h"
@@ -141,6 +142,7 @@ static inline size_t br_port_info_size(void)
 #ifdef CONFIG_BRIDGE_IGMP_SNOOPING
 		+ nla_total_size(sizeof(u8))	/* IFLA_BRPORT_MULTICAST_ROUTER */
 #endif
+		+ nla_total_size(4)	/* IFLA_BRPORT_LEARNING_LIMIT */
 		+ 0;
 }
 
@@ -191,7 +193,8 @@ static int br_port_fill_attrs(struct sk_buff *skb,
 	    nla_put_u16(skb, IFLA_BRPORT_NO, p->port_no) ||
 	    nla_put_u8(skb, IFLA_BRPORT_TOPOLOGY_CHANGE_ACK,
 		       p->topology_change_ack) ||
-	    nla_put_u8(skb, IFLA_BRPORT_CONFIG_PENDING, p->config_pending))
+	    nla_put_u8(skb, IFLA_BRPORT_CONFIG_PENDING, p->config_pending) ||
+	    nla_put_u32(skb, IFLA_BRPORT_LEARNING_LIMIT, p->learning_limit))
 		return -EMSGSIZE;
 
 	timerval = br_timer_value(&p->message_age_timer);
@@ -586,6 +589,7 @@ static const struct nla_policy br_port_policy[IFLA_BRPORT_MAX + 1] = {
 	[IFLA_BRPORT_PROXYARP]	= { .type = NLA_U8 },
 	[IFLA_BRPORT_PROXYARP_WIFI] = { .type = NLA_U8 },
 	[IFLA_BRPORT_MULTICAST_ROUTER] = { .type = NLA_U8 },
+	[IFLA_BRPORT_LEARNING_LIMIT] = { .type = NLA_S32 },
 };
 
 /* Change the state of the port and notify spanning tree */
@@ -610,6 +614,38 @@ static int br_set_port_state(struct net_bridge_port *p, u8 state)
 	return 0;
 }
 
+/* Change the state of the port and notify spanning tree */
+static int br_set_learning_limit(struct net_bridge_port *p, int learning_limit)
+{
+	struct switchdev_attr attr = {
+		.orig_dev = p->dev,
+		.id = SWITCHDEV_ATTR_ID_PORT_LEARNING_LIMIT,
+		.flags = SWITCHDEV_F_SKIP_EOPNOTSUPP,
+		.u.learning_limit = learning_limit,
+	};
+	int err;
+
+	if (learning_limit < BRPORT_LEARNING_LIMIT_UNLIMITED) {
+		br_warn(p->br, "invalid value (%i) for mac table learning on port %u(%s)\n",
+			learning_limit, (unsigned int)p->port_no, p->dev->name);
+		return -EINVAL;
+	}
+
+	p->learning_limit = learning_limit;
+
+	err = switchdev_port_attr_set(p->dev, &attr);
+	if (err && err != -EOPNOTSUPP)
+		br_warn(p->br, "error setting offload mac table learning on port %u(%s)\n",
+			(unsigned int)p->port_no, p->dev->name);
+	else
+		br_debug(p->br, "mac learning limit set to %i on port %u(%s)\n",
+			 p->learning_limit, (unsigned int)p->port_no,
+			 p->dev->name);
+
+	br_configuration_update(p->br);
+	return 0;
+}
+
 /* Set/clear or port flags based on attribute */
 static void br_set_port_flag(struct net_bridge_port *p, struct nlattr *tb[],
 			   int attrtype, unsigned long mask)
@@ -669,6 +705,12 @@ static int br_setport(struct net_bridge_port *p, struct nlattr *tb[])
 			return err;
 	}
 #endif
+	if (tb[IFLA_BRPORT_LEARNING_LIMIT]) {
+		err = br_set_learning_limit(p, nla_get_u32(tb[IFLA_BRPORT_LEARNING_LIMIT]));
+		if (err)
+			return err;
+	}
+
 	br_port_flags_change(p, old_flags ^ p->flags);
 	return 0;
 }
diff --git a/net/bridge/br_private.h b/net/bridge/br_private.h
index 972fc75cf8e4..a58fc166996b 100644
--- a/net/bridge/br_private.h
+++ b/net/bridge/br_private.h
@@ -273,6 +273,7 @@ struct net_bridge_port {
 	bridge_id			designated_bridge;
 	u32				path_cost;
 	u32				designated_cost;
+	int				learning_limit;
 	unsigned long			designated_age;
 
 	struct timer_list		forward_delay_timer;
diff --git a/net/bridge/br_sysfs_if.c b/net/bridge/br_sysfs_if.c
index abf711112418..eb0ea036cea6 100644
--- a/net/bridge/br_sysfs_if.c
+++ b/net/bridge/br_sysfs_if.c
@@ -81,6 +81,13 @@ static ssize_t show_priority(struct net_bridge_port *p, char *buf)
 static BRPORT_ATTR(priority, S_IRUGO | S_IWUSR,
 			 show_priority, br_stp_set_port_priority);
 
+static ssize_t show_learning_limit(struct net_bridge_port *p, char *buf)
+{
+	return sprintf(buf, "%i\n", p->learning_limit);
+}
+
+static BRPORT_ATTR(learning_limit, S_IRUGO, show_learning_limit, NULL);
+
 static ssize_t show_designated_root(struct net_bridge_port *p, char *buf)
 {
 	return br_show_bridge_id(buf, &p->designated_root);
